test/testfloat/testfloat.cpp

```diff
@@ -0,0 +1,442 @@
+{
+    Type m_type;
+    fizzy::Value m_value;
+
+public:
+    constexpr TypedValue(Type type, fizzy::Value value) noexcept : m_type{type}, m_value{value} {}
+
+    constexpr TypedValue(Type type, uint64_t bits) noexcept
+      : m_type{type}, m_value{make_value(type, bits)}
+    {}
+
+    bool eq(uint64_t expected_bits, bool ignore_nans) const
+    {
+        switch (m_type)
+        {
+        case Type::i32:
+        case Type::i64:
+            return m_value.i64 == expected_bits;
+        case Type::f32:
+        {
+            const FP value{m_value.f32};
+            const FP expected{static_cast<uint32_t>(expected_bits)};
+            if (ignore_nans && std::isnan(expected.value))
+            {
+                if (expected.is_canonical_nan())
+                    return value.is_canonical_nan();
+                if (expected.is_arithmetic_nan())
+                    return value.is_arithmetic_nan();
+                throw std::invalid_argument{"invalid input: unexpected signaling NaN"};
+            }
+            return value == expected;
+        }
+
+        case Type::f64:
+        {
+            const FP value{m_value.f64};
+            const FP expected{expected_bits};
+            if (ignore_nans && std::isnan(expected.value))
+            {
+                if (expected.is_canonical_nan())
+                    return value.is_canonical_nan();
+                if (expected.is_arithmetic_nan())
+                    return value.is_arithmetic_nan();
+                throw std::invalid_argument{"invalid input: unexpected signaling NaN"};
+            }
+            return value == expected;
+        }
+        }
+        __builtin_unreachable();
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const TypedValue& x)
+    {
+        std::ios os_state{nullptr};
+        os_state.copyfmt(os);
+        os << std::hex << std::uppercase << std::setfill('0');
+        switch (x.m_type)
+        {
+        case Type::i32:
+            os << std::setw(8) << x.m_value.as<uint32_t>();
+            break;
+        case Type::i64:
+            os << std::setw(16) << x.m_value.i64;
+            break;
+        case Type::f32:
+            os << std::setw(8) << FP{x.m_value.f32}.as_uint();
+            break;
+        case Type::f64:
+            os << std::setw(16) << FP{x.m_value.f64}.as_uint();
+            break;
+        }
+        os.copyfmt(os_state);
+        return os;
+    }
+};
+
+bool check(const FunctionDescription& func, fizzy::Instance& instance, const uint64_t inputs[],
+    bool ignore_nans)
+{
+    const auto report_failure = [&](auto result, auto expected) {
+        std::cerr << "FAILURE: " << result << " <-";
+        for (size_t i = 0; i < func.num_arguments; ++i)
+            std::cerr << ' ' << TypedValue{func.param_types[i], inputs[i]};
+        std::cerr << "\n         " << expected << " (expected)\n";
+    };
+
+    fizzy::Value args[2]{};
+    for (size_t i = 0; i < func.num_arguments; ++i)
```

Comment:
Nit: could add an assert for `num_arguments <= 2`, but I know it is ensured by the declaration in `FunctionDescription`.